runtime/grid_sync.cu

```diff
@@ -75,6 +75,10 @@ __device__ void sync(
       }
 #endif
     }
+    if (last_block) {
+      // Clean up semaphore for re-use in next iteration
+      semaphore = 0ULL;
```

Comment:
We can't do this without causing deadlocks. The problem is that when `PERSISTENT=true` then all blocks will hang around waiting for the last block to finish finish syncing. This is done by watching for the first bit to flip, which is done by the last block. If the last block flips that first bit then immediately resets the semaphore to zero, then those other blocks will see zero instead of the flipped bit which means they will keep waiting indefinitely. It's an interesting way to create a reusable semaphore, but it's unclear to me how we can automatically clean this without just introducing an expression that resets the semaphore at the end of the kernel when we detect it's needed.

Reply:
For now, I think it is actually safe to reset the semaphore after syncing as long as the PERSISTENT flag is false, since this should indicate that the semaphore won't be re-used.